src/yarp.c

```diff
@@ -2283,6 +2225,106 @@ parse_parameters(yp_parser_t *parser) {
   return params;
 }
 
+// Parse a list of arguments and their surrounding parentheses if they are
+// present.
+static void
+parse_arguments_list(yp_parser_t *parser, yp_arguments_t *arguments) {
+  switch (parser->current.type) {
+    case YP_TOKEN_EOF:
+    case YP_TOKEN_BRACE_RIGHT:
+    case YP_TOKEN_BRACKET_RIGHT:
+    case YP_TOKEN_COLON:
+    case YP_TOKEN_COMMA:
+    case YP_TOKEN_EMBEXPR_END:
+    case YP_TOKEN_EQUAL_GREATER:
+    case YP_TOKEN_KEYWORD_DO:
+    case YP_TOKEN_KEYWORD_END:
+    case YP_TOKEN_KEYWORD_IN:
+    case YP_TOKEN_NEWLINE:
+    case YP_TOKEN_PARENTHESIS_RIGHT:
+    case YP_TOKEN_SEMICOLON: {
+      // The reason we need this short-circuit is because we're using the
+      // binding powers table to tell us if the subsequent token could
+      // potentially be an argument. If there _is_ a binding power for one of
+      // these tokens, then we should remove it from this list and let it be
+      // handled by the default case below.
+      assert(binding_powers[parser->current.type].left == BINDING_POWER_UNSET);
+
+      arguments->opening = not_provided(parser);
+      arguments->closing = not_provided(parser);
+      arguments->arguments = NULL;
+
+      break;
+    }
+    case YP_TOKEN_PARENTHESIS_LEFT: {
+      parser_lex(parser);
+      arguments->opening = parser->previous;
+
+      if (accept(parser, YP_TOKEN_PARENTHESIS_RIGHT)) {
+        arguments->arguments = NULL;
+        arguments->closing = parser->previous;
+      } else {
+        arguments->arguments = yp_node_arguments_node_create(parser);
+        parse_arguments(parser, arguments->arguments, YP_TOKEN_PARENTHESIS_RIGHT);
+        expect(parser, YP_TOKEN_PARENTHESIS_RIGHT, "Expected a ')' to close the argument list.");
+        arguments->closing = parser->previous;
+      }
+      break;
+    }
+    default: {
+      arguments->opening = not_provided(parser);
+      arguments->closing = not_provided(parser);
+
+      if (binding_powers[parser->current.type].left == BINDING_POWER_UNSET) {
+        // If we get here, then the subsequent token cannot be used as an infix
+        // operator. In this case we assume the subsequent token is part of an
+        // argument to this method call.
+        arguments->arguments = yp_node_arguments_node_create(parser);
+        parse_arguments(parser, arguments->arguments, YP_TOKEN_EOF);
+      } else {
+        arguments->arguments = NULL;
+      }
+
+      break;
+    }
+  }
+
+  if (accept(parser, YP_TOKEN_BRACE_LEFT) || accept(parser, YP_TOKEN_KEYWORD_DO)) {
+    not_provided(parser);
+    not_provided(parser);
+
+    yp_token_t opening = parser->previous;
+
+    accept(parser, YP_TOKEN_NEWLINE);
+
+    yp_node_t *block_arguments = NULL;
+    if (accept(parser, YP_TOKEN_PIPE)) {
+      block_arguments = parse_parameters(parser);
+      expect(parser, YP_TOKEN_PIPE, "expected block arguements to end with '|'.");
```

Comment:
I didn't handle `accept(parser, YP_TOKEN_SEMICOLON)` here like @paracycle 's change did since I think I would end up duplicating almost exactly what he did there and we should probably share that code somewhere.

Reply:
I'd extracted lambda/block parameter parsing code to its own function in my branch, which you can cherry-pick and start using if you want: https://github.com/Shopify/yarp/pull/217/commits/4a03263cb3f23e2dd55ae3057b8f51adc8f754c4

Reply:
Though I guess you would need this too: https://github.com/Shopify/yarp/pull/217/commits/e9c086dad959d98d694fa30cc4f3bca2967a9fb5#diff-5a7db8dbadaef1b1b5a8738ba70b5ffac82b8e243732154165911284e08aad4bR402-R413

Reply:
Maybe I should wait for lambda support to merge? I think this PR would be ok waiting.